### PR TITLE
feat: Add Transaction/Span Development Docs

### PR DESCRIPTION
**Pull request by HazAT**
First, probably not 100% documentation for Transactions/Spans.

### CODE REVIEW

File: src/docs/sdk/event-payloads/properties/status.mdx [hunk 1]
```diff
@@ -0,0 +1,30 @@
+`status`:
+
+: _Optional_. Describing the `status` of the Span. We use this property to determine if the [Transaction](./transaction)
+ended in an error or not.
+
+State|Description|HTTP status code equivalent
+---|---|---
+`ok`|Not an error, returned on success|200 and 2XX HTTP statuses
+`cancelled`|The operation was cancelled, typically by the caller|499
+`unknown`|An unknown error raised by APIs that don't return enough error information|500
```
Comment (AbhiPrasad) [inline src/docs/sdk/event-payloads/properties/status.mdx:10]: This can also be `unknown_error` -> https://github.com/getsentry/relay/blob/master/relay-general/src/protocol/contexts.rs#L503

Comment (dcramer) [reply src/docs/sdk/event-payloads/properties/status.mdx:10]: probably describe it as "unknown result" vs "unknown error"

File: src/docs/sdk/event-payloads/transaction.mdx [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+title: Transaction Payloads
+---
+
+Transactions are used to send tracing events to Sentry.
+
+Transactions must be wrapped in an [Envelope](/sdk/envelopes) and therefore also be sent to the Envelope endpoint.
+
+<Alert title="Note on backwards compatibility" level="info">
+
+Envelopes are a new format and therefore only work on Sentry >= 10. More infromation can be found on the [Envelope](../envelopes) page.
```
Comment (AbhiPrasad) [inline src/docs/sdk/event-payloads/transaction.mdx:11]:
```suggestion
Envelopes are a new format and therefore only work on Sentry >= 10. More information can be found on the [Envelope](../envelopes) page.
```

File: src/docs/sdk/event-payloads/transaction.mdx [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+</Alert>
+
+## Anatomy
+
+A Transaction is basically a [Span](/sdk/event-payloads/span) combined with an [Event](/sdk/event-payloads/). When using 
+tracing with our SDKs you usually create a Span tree, the root node and therefore the whole tree is considered to be the Transaction.
+So technically a Transaction is just a Span. A Transaction must also have a `contexts.trace` (which contains some data of the [Span](/sdk/event-payloads/span)) and some other
+properties that will be covered in the next section.
+
+Transactions is an [Events](/sdk/event-payloads/) which will be enriched with data from the [Span](/sdk/event-payloads/span).
```
Comment (AbhiPrasad) [inline src/docs/sdk/event-payloads/transaction.mdx:21]:
```suggestion
Transactions are [Events](/sdk/event-payloads/) that will be enriched with data from the [Span](/sdk/event-payloads/span).
```

File: src/docs/sdk/event-payloads/transaction.mdx [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+import "./properties/type.mdx"
+
+import "./properties/start_timestamp.mdx"
+
+import "./properties/timestamp.mdx"
+
+### contexts.trace
+
+A Transaction has to have a specifc `contexts.trace` entry that contains data from the [Span](/sdk/event-payloads/span).
```
Comment (AbhiPrasad) [inline src/docs/sdk/event-payloads/transaction.mdx:32]:
```suggestion
A Transaction has to have a specific `contexts.trace` entry that contains data from the [Span](/sdk/event-payloads/span).
```

File: src/docs/sdk/event-payloads/span.mdx [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+The Span Interface specifies a series of _timed_ application events that have a start and end time.
+
+A [Transaction](./transaction) may contain zero or more Spans in an attribute named `spans`.
+
+This interface is an object with a single `values` attribute containing an
+ordered list of Span objects. Spans in the list don't have to be ordered, they will be ordered
+by start / end time on the Server.
+
+While Span attributes will be normalized on the server, a Span is most
+useful when it includes at least a `op` and `description`.
```
Comment (AbhiPrasad) [inline src/docs/sdk/event-payloads/span.mdx:14]:
```suggestion
useful when it includes at least an `op` and `description`.
```

File: src/docs/sdk/event-payloads/properties/contexts_trace.mdx [hunk 6]
```diff
@@ -0,0 +1,9 @@
+`contexts.trace`:
+
+: _Required_. Contains all the data relevat for the Transaction/Span.
```
Comment (AbhiPrasad) [inline src/docs/sdk/event-payloads/properties/contexts_trace.mdx:3]:
```suggestion
: _Required_. Contains all the data relevant for the Transaction/Span.
```

File: src/docs/sdk/event-payloads/properties/description.mdx [hunk 7]
```diff
@@ -0,0 +1,9 @@
+`description`:
+
+: _Optional_. longer description of the Span's operation, often specific to that instance.
```
Comment (AbhiPrasad) [inline src/docs/sdk/event-payloads/properties/description.mdx:3]:
```suggestion
: _Optional_. Longer description of the Span's operation, often specific to that instance.
```

File: src/docs/sdk/event-payloads/properties/sampled.mdx [hunk 8]
```diff
@@ -0,0 +1,9 @@
+`sampled`:
+
+: _Required_. Determines if the the Span should be send, if `!== true` it will be sampled and therefore not sent.
```
Comment (AbhiPrasad) [inline src/docs/sdk/event-payloads/properties/sampled.mdx:3]:
```suggestion
: _Required_. Determines if the the Span should be sent, if `!== true` it will be sampled and therefore not sent.
```

File: src/docs/sdk/event-payloads/properties/contexts_trace.mdx (hunk 6 above)
Comment (AbhiPrasad) [reply src/docs/sdk/event-payloads/properties/contexts_trace.mdx:3]: What is the purpose of context? Are there fields that are recommended to be included?

File: src/docs/sdk/event-payloads/properties/event_id.mdx [hunk 9]
```diff
@@ -0,0 +1,10 @@
+`event_id`:
+
+: _Required_. Hexadecimal string representing a uuid4 value. The length is exactly 32
```
Comment (AbhiPrasad) [inline src/docs/sdk/event-payloads/properties/event_id.mdx:3]: Just because I'm curious, any reason we opted to use uuid? It might be faster on some platforms to just generate a random `byte` array and encode to string. Ditto with `trace_id` and `span_id`

File: src/docs/sdk/event-payloads/properties/op.mdx [hunk 10]
```diff
@@ -0,0 +1,10 @@
+`op`:
+
+: _Recommended_. Short code identifying the type of operation the span is measuring.
+If no set we fallback to `default` on the Server.
```
Comment (AbhiPrasad) [inline src/docs/sdk/event-payloads/properties/op.mdx:4]:
```suggestion
If no op is set we fallback to `default` on the Server.
```

File: src/docs/sdk/event-payloads/properties/trace_id.mdx [hunk 11]
```diff
@@ -0,0 +1,9 @@
+`trace_id`: 
+
+: _Required_. Determines to wich `trace` the Span belonges to.
```
Comment (AbhiPrasad) [inline src/docs/sdk/event-payloads/properties/trace_id.mdx:3]:
```suggestion
: _Required_. Determines which `trace` the Span belongs to.
```

File: src/docs/sdk/event-payloads/properties/transaction.mdx [hunk 12]
```diff
@@ -0,0 +1,9 @@
+`transaction`: 
+
+: _Optional/Required_. Set the `transaction` to give your tree of spans a dedicated name.
```
Comment (AbhiPrasad) [inline src/docs/sdk/event-payloads/properties/transaction.mdx:3]: What happens if transaction is empty?

File: src/docs/sdk/event-payloads/transaction.mdx [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+Transactions must be wrapped in an [Envelope](/sdk/envelopes) and therefore also be sent to the Envelope endpoint.
+
+<Alert title="Note on backwards compatibility" level="info">
+
+Envelopes are a new format and therefore only work on Sentry >= 10. More infromation can be found on the [Envelope](../envelopes) page.
+</Alert>
+
+## Anatomy
+
+A Transaction is basically a [Span](/sdk/event-payloads/span) combined with an [Event](/sdk/event-payloads/). When using
```
Comment (AbhiPrasad) [inline src/docs/sdk/event-payloads/transaction.mdx:16]: Are there fields from `Event` that have no use if set on a `Transaction`?

File: src/docs/sdk/event-payloads/transaction.mdx [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+
+```json
+{
+    "contexts": {
+        "trace": {
+            "op": "navigation",
+            "description": "User clicked on <Link />",
+            "trace_id": "743ad8bbfdd84e99bc38b4729e2864de",
+            "span_id": "a0cfbde2bdff3adc"
+        }
```
Comment (rhcarvalho) [inline src/docs/sdk/event-payloads/transaction.mdx:54]: Need to review/align this with code in Relay/SDKs.

---

I looked at an existing transaction sent from the JS SDK and it has an additional `"type": "trace"`.

```json
    "contexts": {
        "trace": {
            "description": "[redacted]",
            "op": "[redacted]",
            "type": "trace",
            "trace_id": "f0eeddd8e915418b953ff969df526b2d",
            "span_id": "96f3b555b5c7ad45"
        }
    },
```

And looked at Relay: https://github.com/getsentry/relay/blob/53a260e6de70d1fec496aaf8c9348f0c5a17e5d8/relay-general/src/protocol/contexts.rs#L360-L383

- Relay ignores the `type` key (we should fix it in the SDK)
- Relay ignores the `description` (goes to `other` field)
- Relay knows of a `status` which is not present here
- Relay knows of a `parent_span_id`, also not here

File: src/docs/sdk/event-payloads/properties/contexts_trace.mdx (hunk 6 above)
Comment (HazAT) [reply src/docs/sdk/event-payloads/properties/contexts_trace.mdx:3]: Yeah they have to be there and "represent" the Span

File: src/docs/sdk/event-payloads/properties/sampled.mdx (hunk 8 above)
[the remark this answers is not in the document]
Comment (HazAT) [reply src/docs/sdk/event-payloads/properties/sampled.mdx:3]: Yeah, I would say it's up to the SDK to decide, relay shouldn't reject this.

File: src/docs/sdk/event-payloads/transaction.mdx (hunk 13 above)
Comment (HazAT) [reply src/docs/sdk/event-payloads/transaction.mdx:16]: Kinda, so a transaction should not have a stacktrace/exception/threads but we can take care of this in the next iteration.
Not that it will break it's just not defined what happens.

File: src/docs/sdk/event-payloads/properties/sampled.mdx (hunk 8 above)
[the remark this answers is not in the document]
Comment (rhcarvalho) [reply src/docs/sdk/event-payloads/properties/sampled.mdx:3]: Since now we are finally documenting what the protocol is, what is the role of sending `sampled: true` or `sampled: false` and marking it as **required** field?

File: src/components/sidebar.js [hunk 15]
```diff
@@ -120,6 +121,8 @@ const Sidebar = () => (
         <NavLink to="/sdk/store/">Store Endpoint</NavLink>
         <NavLink to="/sdk/envelopes/">Envelopes</NavLink>
         <NavLink to="/sdk/event-payloads/" title="Event Payloads">
+          <NavLink to="/sdk/event-payloads/transaction/">Transaction Type</NavLink>
+          <NavLink to="/sdk/event-payloads/span/">Span Interface</NavLink>
```
Comment (rhcarvalho) [inline src/components/sidebar.js:125]: Looking at how this will be rendered, it is strange to have Transaction and Span above all other Interfaces that are part of regular (error) events.

_Since this is related to the bigger difficulty in explaining Events x Transactions and what are valid attributes in each, feel free to ignore me..._

File: src/components/sidebar.js [hunk 16]
```diff
@@ -4,13 +4,14 @@ import { useLocation } from "@reach/router";
...
-const NavLink = ({ to, title, children }) => {
+const NavLink = ({ to, title, children, ...props }) => {
   const location = useLocation();
-
+  
   let className = "toc-item";
   if (location && location.pathname.indexOf(withPrefix(to)) === 0) {
     className += " toc-active";
   }
+  className += (props.className ? ' ' + props.className : '');
```
Comment (rhcarvalho) [inline src/components/sidebar.js:14]: How is this related to the rest of the PR? Is it used?

File: src/docs/sdk/event-payloads/properties/op.mdx (hunk 10 above)
Comment (rhcarvalho) [reply src/docs/sdk/event-payloads/properties/op.mdx:4]: This was marked as resolved without applying the change, assuming it was an accident.

TBH, I'd rather we give guidance on `op` values than saying what Relay does (I consider that an implementation detail). I wouldn't mind if we removed this sentence altogether.

File: src/docs/sdk/event-payloads/properties/sampled.mdx (hunk 8 above)
[the remark this answers is not in the document]
Comment (rhcarvalho) [reply src/docs/sdk/event-payloads/properties/sampled.mdx:3]: @HazAT I believe `sampled` should not be part of the JSON payload.

File: src/docs/sdk/event-payloads/transaction.mdx [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+
+import "./properties/type.mdx"
+
+import "./properties/span_start_timestamp.mdx"
+
+import "./properties/span_timestamp.mdx"
+
+### contexts.trace
+
+A Transaction has to have a specific `contexts.trace` entry that contains data from the [Span](/sdk/event-payloads/span).
```
Comment (rhcarvalho) [inline src/docs/sdk/event-payloads/transaction.mdx:32]: I believe this should be documented similar to other `contexts`, specially because we're sending `contexts.trace` also in error events / `captureMessage`.

(I've asked about this on Slack earlier this week after seeing `contexts.trace` in the payload from `captureMessage`)

File: src/docs/sdk/event-payloads/properties/trace_id.mdx [hunk 18]
```diff
@@ -0,0 +1,9 @@
+`trace_id`: 
+
+: _Required_. Determines to which `trace` the Span belonges to.
```
Comment (rhcarvalho) [inline src/docs/sdk/event-payloads/properties/trace_id.mdx:3]: (copying from an outdated suggestion from @AbhiPrasad + my previous suggestion to document the content):

```suggestion
: _Required_. Determines which `trace` the Span belongs to. The value should be 16 random bytes encoded as a hex string (32 characters long).
```

File: src/docs/sdk/event-payloads/properties/transaction.mdx [hunk 19]
```diff
@@ -0,0 +1,9 @@
+`transaction`: 
+
+: _Recommended_. In case of a root span, this property is recommended. Set the `transaction` to give your tree of spans a dedicated name. If it's empty we fall back to `<unlabeled transaction>`.
```
Comment (rhcarvalho) [inline src/docs/sdk/event-payloads/properties/transaction.mdx:3]: Is it valid for non-root spans? Is there a defined behavior in that case?

Like for `op`, I feel the exact behavior in Relay is an implementation detail, I'd leave it out. Your call.

File: src/docs/sdk/event-payloads/span.mdx [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+
+The following example illustrates the Span as part of the [Transaction](./transaction) and omits other attributes for simplicity.
+
+```json
+{
+  "spans": [
+    {
+      "sampled": true,
+      "start_timestamp": 1588601261.481961,
+      "description": "GET http://localhost:8080/sockjs-node/info?t=1588601703755",
```
Comment (rhcarvalho) [inline src/docs/sdk/event-payloads/span.mdx:65]: To be consistent with the docs for `description`:

```suggestion
      "description": "GET /sockjs-node/info",
```

File: src/docs/sdk/event-payloads/properties/span_start_timestamp.mdx [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+: _Required_. A timestamp representing when the measuring started. The
+  format is either a string as defined in [RFC
+  3339](https://tools.ietf.org/html/rfc3339) or a numeric (integer or float)
+  value representing the number of seconds that have elapsed since the [Unix
+  epoch](https://en.wikipedia.org/wiki/Unix_time). Important to note is that the `start_timestamp`
+  must be before the `timestamp` otherwise we will discard the Span as invalid.
+
+  ```json
+  {
+    "start_timestamp": "2011-05-02T17:41:36Z"
```
Comment (rhcarvalho) [inline src/docs/sdk/event-payloads/properties/span_start_timestamp.mdx:12]: Some changes got lost when renaming the file...

```suggestion
    "start_timestamp": "2011-05-02T17:41:36.242Z"
```

File: src/docs/sdk/event-payloads/properties/span_start_timestamp.mdx [hunk 22]
```diff
@@ -0,0 +1,22 @@
...
+  }
+  ```
+
+or: 
+
+  ```json
+  {
+    "start_timestamp": 1304358096.0
```
Comment (rhcarvalho) [inline src/docs/sdk/event-payloads/properties/span_start_timestamp.mdx:20]:
```suggestion
    "start_timestamp": 1304358096.242
```

File: src/docs/sdk/event-payloads/properties/span_start_timestamp.mdx [hunk 23]
```diff
@@ -0,0 +1,22 @@
+`start_timestamp`:
+
+: _Required_. A timestamp representing when the measuring started. The
+  format is either a string as defined in [RFC
+  3339](https://tools.ietf.org/html/rfc3339) or a numeric (integer or float)
+  value representing the number of seconds that have elapsed since the [Unix
+  epoch](https://en.wikipedia.org/wiki/Unix_time). Important to note is that the `start_timestamp`
+  must be before the `timestamp` otherwise we will discard the Span as invalid.
```
Comment (rhcarvalho) [inline src/docs/sdk/event-payloads/properties/span_start_timestamp.mdx:8]:
```suggestion
  epoch](https://en.wikipedia.org/wiki/Unix_time). The `start_timestamp` value must be greater or
  equal the `timestamp` value, otherwise the Span is discarded as invalid.
```

It is valid to have start == end.

File: src/docs/sdk/event-payloads/properties/span_timestamp.mdx [hunk 24]
```diff
@@ -0,0 +1,21 @@
...
+
+: _Required_. A timestamp representing when the measuring finished. The
+  format is either a string as defined in [RFC
+  3339](https://tools.ietf.org/html/rfc3339) or a numeric (integer or float)
+  value representing the number of seconds that have elapsed since the [Unix
+  epoch](https://en.wikipedia.org/wiki/Unix_time).
+
+  ```json
+  {
+    "timestamp": "2011-05-02T17:41:36Z"
```
Comment (rhcarvalho) [inline src/docs/sdk/event-payloads/properties/span_timestamp.mdx:11]:
```suggestion
    "timestamp": "2011-05-02T17:41:36.955Z"
```

File: src/docs/sdk/event-payloads/properties/span_timestamp.mdx [hunk 25]
```diff
@@ -0,0 +1,21 @@
...
+  }
+  ```
+
+or: 
+
+  ```json
+  {
+    "timestamp": 1304358096.0
```
Comment (rhcarvalho) [inline src/docs/sdk/event-payloads/properties/span_timestamp.mdx:19]:
```suggestion
    "timestamp": 1304358096.955
```

File: src/docs/sdk/event-payloads/properties/spans.mdx [hunk 26]
```diff
@@ -0,0 +1,39 @@
...
+         "start_timestamp":1588601261.535386,
+         "description":"Vue <App>",
+         "timestamp":1588601261.544196,
+         "parent_span_id":"9312d0d18bf51736",
+         "trace_id":"1e57b752bc6e4544bbaa246cd1d05dee",
+         "op":"update",
+         "span_id":"b980d4dec78d7344"
+      }
+   ]
+}
```
Comment (rhcarvalho) [inline src/docs/sdk/event-payloads/properties/spans.mdx:38]:
```suggestion
{
  "spans": [
    {
      "start_timestamp": 1588601261.481961,
      "description": "GET /sockjs-node/info",
      "tags": {
        "http.status_code": "200"
      },
      "timestamp": 1588601261.488901,
      "parent_span_id": "b0e6f15b45c36b12",
      "trace_id": "1e57b752bc6e4544bbaa246cd1d05dee",
      "op": "http",
      "data": {
        "url": "http://localhost:8080/sockjs-node/info?t=1588601703755",
        "status_code": 200,
        "type": "xhr",
        "method": "GET"
      },
      "span_id": "b01b9f6349558cd1"
    },
    {
      "start_timestamp": 1588601261.535386,
      "description": "Vue <App>",
      "timestamp": 1588601261.544196,
      "parent_span_id": "9312d0d18bf51736",
      "trace_id": "1e57b752bc6e4544bbaa246cd1d05dee",
      "op": "update",
      "span_id": "b980d4dec78d7344"
    }
  ]
}
```

File: src/components/sidebar.js (hunk 16 above)
Comment (HazAT) [reply src/components/sidebar.js:14]: Experimented with the sidebar and classes. Now you cant set classname on a navigation item but not related to this PR. Still like to leave it.
